go/vt/vtgate/planbuilder/rewrite.go

```diff
@@ -109,48 +107,96 @@ func expandTableColumns(tables []semantics.TableInfo, starExpr *sqlparser.StarEx
 	return starExpanded, colNames, nil
 }
 
-func subqueryRewrite(ctx planningContext, statement sqlparser.SelectStatement) error {
-	if len(ctx.semTable.SubqueryMap) == 0 {
-		return nil
-	}
+func queryRewrite(ctx planningContext, statement sqlparser.SelectStatement) error {
 	r := rewriter{
 		semTable:     ctx.semTable,
 		reservedVars: ctx.reservedVars,
 	}
-	sqlparser.Rewrite(statement, r.subqueryRewrite, nil)
+	sqlparser.Rewrite(statement, r.rewrite, nil)
 	return nil
 }
 
-func (r *rewriter) subqueryRewrite(cursor *sqlparser.Cursor) bool {
+func (r *rewriter) rewrite(cursor *sqlparser.Cursor) bool {
 	switch node := cursor.Node().(type) {
+	case *sqlparser.Select:
+		rewriteHavingClause(node)
 	case *sqlparser.ExistsExpr:
-		semTableSQ, found := r.semTable.SubqueryRef[node.Subquery]
-		if !found {
-			// should never happen
-			return false
-		}
+		return r.rewriteExistsSubquery(cursor, node)
+	case *sqlparser.Subquery:
+		rewriteSubquery(cursor, r, node)
```

Comment:
I think we should either have both functions be methods of `rewriter`, or have them both take `r rewriter` as function parameter

Reply:
I fix in the next PR